test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,307 @@
+	osType   string
+	ifName   string
+	ip       string
+	ifIndex  string // Used only for Windows
+}
+
+// TestVMAgent is the top-level test which can contain some subtests for
+// VMAgent so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfFeatureDisabled(t, features.ExternalNode, false, true)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	vmList, err := setupVMAgentTest(t, data)
+	if err != nil {
+		t.Fatalf("Error when setting up VMAgent test: %v", err)
+	}
+	defer teardownVMAgentTest(t, data, vmList)
+	t.Run("testExternalNode", func(t *testing.T) { testExternalNode(t, data, vmList) })
+}
+
+// setupVMAgentTest creates ExternalNode, starts antrea-agent
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.winVMs, testOptions.linVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), nameSpace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(nameSpace).Get(context.TODO(), serviceAccount, metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using namespace %s", nsReturned.Name)
+	var vmList []vmInfo
+	if testOptions.linVMs != "" {
+		vms := strings.Split(testOptions.linVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	if testOptions.winVMs != "" {
+		vms := strings.Split(testOptions.winVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM: %s", vm)
+			tempVm := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	t.Logf("TestVMAgent setup")
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for VM: %s", vm.nodeName)
+		_, err := createExternalNodeCRD(data, vm.nodeName, vm.ifName, vm.ip)
+		require.Nil(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
+			if vm.osType == "Linux" {
+				tempVm = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVm.ifName {
+				t.Errorf("Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+				return false, nil
+			}
+			if vm.ip != tempVm.ip {
+				t.Errorf("Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+				return false, nil
+			}
```

Comment:
sometimes you use testify assertions, and sometimes you use an if statement and `t.Error`. Why not be consistent? You can use `assert.Equal` here for example

Reply:
These checks are inside a wait.PollImmediate(), which we expect to return true within a specified timeout. 
That is the reason, we are not have ing assert check in this block.

Reply:
I don't understand your comment. There is no functional difference between using `t.Error` and `assert`. This code is equivalent to:
```go
			if !assert.Equal(t, vm.ifName, tempVm.ifName, "Unexpected uplink interface name") {
				return false, nil
			}
			if !assert.Equal(t, vm.ip, tempVm.ip, ""Unexpected uplink interface IP") {
				return false, nil
			}
```

Reply:
I did not know both are same. I will incorporate your suggestion.